solutions/gold/usaco-814.mdx

```diff
@@ -0,0 +1,155 @@
+---
+id: usaco-814
+source: USACO Gold 2018 February
+title: Directory Traversal
+author: Ryan Chou
+---
+
+<Spoiler title="Hint 1">
+
+If you're not sure where to start, think about what the directories and files represent. Then, you should be able to restructure the problem into one that seems easier to approach.
+
+</Spoiler>
+
+<Spoiler title="Answer to Hint 1">
+
+They form a tree where the lengths of the file/directory names represent the weights.
+
+Now, we can phrase the problem as follows:
+
+> For each node in a tree, what's the minimum sum of path lengths from a node to all of the tree's leaves?
+
+</Spoiler>
+
+<Spoiler title="Solution">
+
+## Explanation
+
+To compute the distances from each node, we'll calculate the "inside" and "outside" distances for each node.
+
+The "inside" distance is the sum of path lengths from a node to the leaves **inside** its subtree.
+
+The "outside" distance is the sum of path lengths from a node to the leaves **outside** its subtree.
+
+For example, in the following tree, the inside distance for node $4$ would equal $4$, and the outside distance would equal $7$.
+
+![](usaco-814/diagram.png)
+
+*Note: For the outside value, we also have to include the two `../`s that are required to backtrack.*
+
+<Optional title="Keep Thinking!">
+
+Now that you've made it here, try figuring out a way to calculate this by yourself. :)
+
+</Optional>
+
+<Spoiler title="Calculating inside distances">
+
+We can solve this with a [DFS](/silver/graph-traversal) through the tree since the inside value of the parent is just the inside value of the child (plus a slash).
+
+</Spoiler>
+
+<Spoiler title="Calculating outside distances">
+
+Surprisingly, we can also solve this with a DFS!
+
+Note that if we have already calculated the outside value for a node's parent, we have all the information we need to calculate itself as well.
+
+This is because every outside node must be in the parent's subtree ($\texttt{inside}[p] - \texttt{inside}[n]$), or outside that parent's subtree ($\texttt{outside}[p]$). Here, $p$ represents the parent, and $n$ represents the current node.
+
+![](usaco-814/outside.png)
+
+</Spoiler>
+
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(N)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+struct Data {
+	bool isFile;  // is it a leaf
```

Comment:
snake_case

Reply:
i have never done that before i don't know what's gotten into me.